test/testbufferoverrun.cpp

```diff
@@ -4197,6 +4199,41 @@ class TestBufferOverrun : public TestFixture {
             "}");
         ASSERT_EQUALS("[test.cpp:4] -> [test.cpp:1]: (error) Pointer arithmetic overflow; 'p' buffer size is 12\n", errout.str());
     }
+
+    void objectIndex()
+    {
+        check("int f() { \n"
+              "    int i;\n"
+              "    return (&i)[1]; \n"
+              "}\n");
+        ASSERT_EQUALS(
+            "[test.cpp:3] -> [test.cpp:3]: (error) The address of variable 'i' is an array of size 1, but the index is non-zero.\n",
```

Comment:
It does not seem obvious that 2 error locations are needed (seems to be the same). Could you call `setMultiline()` at the start of this function so we get more details?

Reply:
thinking loudly; I wonder what you think about changing "but the index is non-zero" to "but the array index is non-zero"

I somehow don't like "The address of variable 'i' is an array of size 1". Not sure how we write that properly. I think it's possible to merge this but well if you have an idea to improve it..

Reply:
One is for the index operator and the other is for the address of operator, which comes from lifetime value:

```
6821.cpp:3:16: warning: The address of variable 'i' is an array of size 1, but the index is non-zero. [objectIndex]
    return (&i)[1];
               ^
6821.cpp:3:13: note: Address of variable taken here.
    return (&i)[1];
            ^
6821.cpp:3:16: note: The address of variable 'i' is an array of size 1, but the index is non-zero.
    return (&i)[1];
               ^
```

It is somewhat trivial for this case, but it can deduce this from many steps, like:

```cpp
int f() { 
    int i = 0;
    int& j = i;
    int* jp = &j;
    return jp[1]; 
}
```

which shows:

```
6821.cpp:5:14: warning: The address of variable 'i' is an array of size 1, but the index is non-zero. [objectIndex]
    return jp[1];
             ^
6821.cpp:3:12: note: Assigned to reference.
    int& j = i;
           ^
6821.cpp:4:15: note: Address of variable taken here.
    int* jp = &j;
              ^
6821.cpp:5:14: note: The address of variable 'i' is an array of size 1, but the index is non-zero.
    return jp[1];
             ^
```

Reply:
Let me try to think of a better message.

Reply:
> It is somewhat trivial for this case,

I wonder if we can avoid it in such trivial case. Something like:

     if (v->errorPath.size() == 1 && v->errorPath.tok() == tok)
        ... skip v->errorPath ...

>  but it can deduce this from many steps, like:

I like that!

Reply:
Doing `if (v->errorPath.size() == 1 && v->errorPath.tok() == tok)` wouldn't fix this issue. Because `tok` points to the `[` token, and the `errorPath.tok()` points to `&`, so it will be false;

Reply:
ok.. so `if (v->errorPath.size() == 1 && v->errorPath.tok() == tok->astOperand1())`?